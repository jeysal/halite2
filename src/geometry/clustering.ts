import { DBSCAN } from 'density-clustering';
import { Point } from 'halite/Geometry';

export const dbscan = <T extends Point>(
  points: T[],
  epsilon: number,
  minPts: number,
): T[][] =>
  new DBSCAN()
    .run(points.map(point => [point.x, point.y]), epsilon, minPts)
    .map(cluster => cluster.map(i => points[i]));
