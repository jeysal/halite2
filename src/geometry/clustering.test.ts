import { dbscan } from './clustering';

describe('dbscan', () => {
  test('clusters points correctly', () => {
    expect(
      dbscan(
        [
          { x: 0, y: 0.5 },
          { x: 3, y: 1.2 },
          { x: -2, y: 3 },
          { x: 0.7, y: 1 },
          { x: 7.5, y: -10 },
          { x: 8.3, y: -8 },
          { x: 5, y: -10 },
          { x: 10, y: 1.1 },
          { x: 9.4, y: -1 },
        ],
        5,
        3,
      ),
    ).toMatchSnapshot();
  });
});
