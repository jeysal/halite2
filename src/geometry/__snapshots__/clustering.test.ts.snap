// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dbscan clusters points correctly 1`] = `
Array [
  Array [
    Object {
      "x": 0,
      "y": 0.5,
    },
    Object {
      "x": 3,
      "y": 1.2,
    },
    Object {
      "x": -2,
      "y": 3,
    },
    Object {
      "x": 0.7,
      "y": 1,
    },
  ],
  Array [
    Object {
      "x": 7.5,
      "y": -10,
    },
    Object {
      "x": 8.3,
      "y": -8,
    },
    Object {
      "x": 5,
      "y": -10,
    },
  ],
]
`;
